network/p2p/topic_validator_test.go

```diff
@@ -109,31 +109,328 @@ func TestTopicValidator(t *testing.T) {
 	err = unstakedNode.Publish(timedCtx, badTopic, data2)
 	require.NoError(t, err)
 
-	// it receives its own message
-	msg, err = unstakedSub.Next(timedCtx)
-	require.NoError(t, err)
-	require.Equal(t, msg.Data, data2)
+	// unstaked node receives its own message
+	checkReceive(timedCtx, t, data2, unstakedSub, nil, true)
 
 	// node 1 does NOT receive the message due to the topic validator
 	var wg sync.WaitGroup
-	wg.Add(1)
+
 	timedCtx, cancel := context.WithTimeout(context.Background(), time.Second)
 	defer cancel()
-	go func() {
-		msg, err = sub1.Next(timedCtx)
-		require.Error(t, err)
-		wg.Done()
-	}()
+	checkReceive(timedCtx, t, nil, sub1, &wg, false)
 
 	// node 2 also does not receive the message via gossip from the node1 (event after the 1 second hearbeat)
-	wg.Add(1)
 	timedCtx, cancel = context.WithTimeout(context.Background(), 2*time.Second)
 	defer cancel()
-	go func() {
-		msg, err = sub2.Next(timedCtx)
-		require.Error(t, err)
-		wg.Done()
-	}()
+	checkReceive(timedCtx, t, nil, sub2, &wg, false)
+
+	unittest.RequireReturnsBefore(t, wg.Wait, 5*time.Second, "could not receive message on time")
+}
+
+// TestAuthorizedSenderValidator_Authorized tests that the authorized sender validator rejects messages from nodes that are not authorized to send the message
+func TestAuthorizedSenderValidator_UnAuthorized(t *testing.T) {
```

Comment:
This tests that a node is not authorized because they have the wrong role. Can you add a test for the case where they are not staked.

Can you also add a test for the case where the message type was sent in the wrong channel

Reply:
Already done [here](https://github.com/onflow/flow-go/pull/2317/files#diff-f818a0f175186a5b2b5e837819b1f1fe7c16d9fe5f24c989100e4eb7e07ab7c3R239)

Reply:
Does that condition get exercised in the test. I thought that test included all of the identities in the identity list, so none of the message would come from an unknown/unstaked node.

Reply:
Yes see [here](https://github.com/onflow/flow-go/blob/f3ce2727c2a0423da7bfaefd4137e30c48f22fbe/network/p2p/topic_validator_test.go#L146) we are trying to broadcast a block proposal message on the sync committee channel. This could be improved by adding a regex on the log statement instead of relying on the log counter only.

Reply:
sorry, I missed that. Should have read the comments. 

My last comment was about the not staked test. Your link pointed to line 239, but it looks like that conditional wouldn't match in the test. maybe I'm being dense

Reply:
Sorry I totally missed that in your previous message i'll add that test case (unstaked node cannot send)